### PR TITLE
[MXPlayer] Rewrite Extractor

### PR DESCRIPTION
**Pull request by Ashish0804**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Allows downloading episodes from mxplayer with the episode url.

### CODE REVIEW

File: yt_dlp/extractor/mxplayer.py [hunk 1]
```diff
@@ -28,12 +28,41 @@ class MxplayerIE(InfoExtractor):
...
+            message = 'No stream provider found'
+            raise ExtractorError('%s said: %s' % (self.IE_NAME, message), expected=True)
+
+        streams = []
+        if stream_provider_dict['dashUrl']:
+            streams.append(('dash', stream_provider_dict['dashUrl']))
+        if stream_provider_dict['hlsUrl']:
+            streams.append(('hls', stream_provider_dict['hlsUrl']))
+        return streams
+
```
Comment (pukkandan) [inline yt_dlp/extractor/mxplayer.py:65]: A lot of code reused from `_get_stream_urls`. Generalize into a single function

File: yt_dlp/extractor/mxplayer.py [hunk 2]
```diff
@@ -69,27 +98,56 @@ def _real_extract(self, url):
...
+                elif stream_name == 'hls':
+                    formats.extend(self._extract_m3u8_formats(stream_url, video_id, fatal=False))
+            info = {
+                'id': video_id,
+                'title': title,
+                'description': video_dict.get('description'),
+                'formats': formats,
+                'season': season,
+                'series': series
+            }
```
Comment (pukkandan) [inline yt_dlp/extractor/mxplayer.py:150]: A lot of code is same as above. Generalize

File: yt_dlp/extractor/mxplayer.py [hunk 3]
```diff
@@ -69,27 +98,56 @@ def _real_extract(self, url):
...
+                if stream_name == 'dash':
+                    format_url = url_or_none(urljoin(config_dict['videoCdnBaseUrl'], stream_url))
+                    if not format_url:
+                        continue
+                    formats.extend(self._extract_mpd_formats(
+                        format_url, video_id, mpd_id='dash', headers=headers))
+                elif stream_name == 'hls':
+                    if stream_url[:4] == 'https':
+                        formats.extend(self._extract_m3u8_formats(stream_url, video_id, fatal=False))
+                    else:
```
Comment (pukkandan) [inline yt_dlp/extractor/mxplayer.py:117]: This case is handled by `urljoin` itself

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/mxplayer.py (hunk 1 above)
Comment (Ashish0804) [reply yt_dlp/extractor/mxplayer.py:65]: only the lines 52,53 and 55,56,57 are same.

Movies and shows serve the url differently.
Movies serve them as a dict with 'hls' or 'dash' name whereas with shows, it's just a key value pair.

Comment (pukkandan) [reply yt_dlp/extractor/mxplayer.py:65]: So only L60-63 are different. You can pass the `type` into the function and easily put this in a if condition

File: yt_dlp/extractor/mxplayer.py [hunk 4]
```diff
@@ -69,27 +98,53 @@ def _real_extract(self, url):
...
+            stream_urls = self._get_stream_urls(video_dict)
+            formats = []
+            headers = {'Referer': url}
+            for stream_name, stream_url in stream_urls:
+                if stream_name == 'dash':
+                    format_url = url_or_none(urljoin(config_dict['videoCdnBaseUrl'], stream_url))
+                    if not format_url:
+                        continue
+                    formats.extend(self._extract_mpd_formats(
+                        format_url, video_id, mpd_id='dash', headers=headers))
```
Comment (pukkandan) [inline yt_dlp/extractor/mxplayer.py:113]:
```suggestion
                    if format_url:
                        formats.extend(self._extract_mpd_formats(
                            format_url, video_id, mpd_id='dash', headers=headers))
```

may be more readable. Just a suggestion, not a needed change

File: yt_dlp/extractor/mxplayer.py (hunk 1 above)
Comment (Ashish0804) [reply yt_dlp/extractor/mxplayer.py:65]: got it

File: yt_dlp/extractor/mxplayer.py [hunk 5]
```diff
@@ -69,27 +98,52 @@ def _real_extract(self, url):
...
+                if stream_name == 'dash':
+                    format_url = url_or_none(urljoin(config_dict['videoCdnBaseUrl'], stream_url))
+                    if format_url:
+                        formats.extend(self._extract_mpd_formats(
+                            format_url, video_id, mpd_id='dash', headers=headers))
+                elif stream_name == 'hls':
+                    format_url = url_or_none(urljoin(config_dict['videoCdnBaseUrl'], stream_url))
+                    if not format_url:
+                        continue
+                    formats.extend(self._extract_m3u8_formats(format_url, video_id, fatal=False))
```
Comment (pukkandan) [inline yt_dlp/extractor/mxplayer.py:117]:
```suggestion
                    if not format_url:
                        formats.extend(self._extract_m3u8_formats(format_url, video_id, fatal=False))
```
same as above

Also, are headers not needed here?

**Review by pukkandan — APPROVED** [review]

**Pull request merged**